# FIFA '21 Messy Raw Dataset Cleaning Challenge

### Objectives of the Data Cleaning Process

1. Ensure that all the columns have the correct data type.
2. Numerical columns should be in a format suitable for further calculations and analysis.

### Getting Started -  Import the Libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Load data
fifa_df = pd.read_csv('../DataChallenge/fifa21_raw_data_v2.csv', low_memory = False, index_col = 'ID')

## 1. Examine the Data Rows and Columns
Now that the data is loaded, let's take a closer Look:
- How many data points do we have?
- How many attributes are there?
- We can get the dimensions of our DataFrame using `df.shape`.

In [3]:
# Investigate shape
fifa_df.shape

(18979, 76)

In [4]:
# Print rows and columns
print("Number of rows :", fifa_df.shape[0])
print("Number of columns :", fifa_df.shape[1])

Number of rows : 18979
Number of columns : 76


In [5]:
# Obtain information about the column data types
fifa_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 76 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Name              18979 non-null  object
 1   LongName          18979 non-null  object
 2   photoUrl          18979 non-null  object
 3   playerUrl         18979 non-null  object
 4   Nationality       18979 non-null  object
 5   Age               18979 non-null  int64 
 6   ↓OVA              18979 non-null  int64 
 7   POT               18979 non-null  int64 
 8   Club              18979 non-null  object
 9   Contract          18979 non-null  object
 10  Positions         18979 non-null  object
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  Preferred Foot    18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  Best Position     18979 non-null  object
 16  Joined            18979 non-null  object
 17  Loan D

We can now see that the dataset consists of 2 data types:
- int64 (whole number),
- object (string or non-numeric).

We can only observe that 'Loan Date End' and 'Hits' column consists of some missing values. But before then, lets get to see the fifa dataframe first five rows.

In [6]:
# Get fifa_df first five rows
fifa_df.head()

Name                      LongName  \
ID                                                        
158023           L. Messi                  Lionel Messi   
20801   Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
200389           J. Oblak                     Jan Oblak   
192985       K. De Bruyne               Kevin De Bruyne   
190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                                photoUrl  \
ID                                                         
158023  https://cdn.sofifa.com/players/158/023/21_60.png   
20801   https://cdn.sofifa.com/players/020/801/21_60.png   
200389  https://cdn.sofifa.com/players/200/389/21_60.png   
192985  https://cdn.sofifa.com/players/192/985/21_60.png   
190871  https://cdn.sofifa.com/players/190/871/21_60.png   

                                                playerUrl Nationality  Age  \
ID                                                                           
158023  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
20801   http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
200389  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
192985  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
190871  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   

        ↓OVA  POT                         Club     Contract  ...     A/W  \
ID                                                           ...           
158023    93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  ...  Medium   
20801     92   92             \n\n\n\nJuventus  2018 ~ 2022  ...    High   
200389    91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023  ...  Medium   
192985    91   91      \n\n\n\nManchester City  2015 ~ 2023  ...    High   
190871    91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022  ...    High   

           D/W   IR PAC  SHO PAS DRI DEF PHY Hits  
ID                                                 
158023     Low  5 ★  85   92  91  95  38  65  771  
20801      Low  5 ★  89   93  81  89  35  77  562  
200389  Medium  3 ★  87   92  78  90  52  90  150  
192985    High  4 ★  76   86  93  88  64  78  207  
190871  Medium  5 ★  91   85  86  94  36  59  595  

[5 rows x 76 columns]

## 2. Cleaning the Data
## 'Name', 'LongName' and 'PlayerUrl'
Straight away there are some interesting things,

'Name', 'LongName' and 'PlayerUrl' column had all information concerning the player names and also several non letter characters, but only the 'playerUrl' column had all the player full names embedded within the Url.

In [7]:
# Inspect playerUrl column
fifa_df['playerUrl'].head()

ID
158023    http://sofifa.com/player/158023/lionel-messi/2...
20801     http://sofifa.com/player/20801/c-ronaldo-dos-s...
200389    http://sofifa.com/player/200389/jan-oblak/210006/
192985    http://sofifa.com/player/192985/kevin-de-bruyn...
190871    http://sofifa.com/player/190871/neymar-da-silv...
Name: playerUrl, dtype: object

- We can then get the players Full Name from the 'playerUrl' column by removing unwanted characters from the column.

In [8]:
# Step 1
fifa_df['playerUrl'] = fifa_df['playerUrl'].str.slice(31)
fifa_df['playerUrl']

ID
158023                  /lionel-messi/210006/
20801     c-ronaldo-dos-santos-aveiro/210006/
200389                     /jan-oblak/210006/
192985               /kevin-de-bruyne/210006/
190871     /neymar-da-silva-santos-jr/210006/
                         ...                 
247223                        /ao-xia/210006/
258760                     /ben-hough/210006/
252757                /ronan-mckinley/210006/
243790                   /zhenao-wang/210006/
252520                     /xiao-zhou/210006/
Name: playerUrl, Length: 18979, dtype: object

In [9]:
# Step 2
fifa_df['playerUrl'] = fifa_df['playerUrl'].str.lstrip('/')
fifa_df['playerUrl']

ID
158023                   lionel-messi/210006/
20801     c-ronaldo-dos-santos-aveiro/210006/
200389                      jan-oblak/210006/
192985                kevin-de-bruyne/210006/
190871      neymar-da-silva-santos-jr/210006/
                         ...                 
247223                         ao-xia/210006/
258760                      ben-hough/210006/
252757                 ronan-mckinley/210006/
243790                    zhenao-wang/210006/
252520                      xiao-zhou/210006/
Name: playerUrl, Length: 18979, dtype: object

In [10]:
# Step 3
fifa_df['playerUrl'] = fifa_df['playerUrl'].str.replace('\d+', '', regex = True)
fifa_df['playerUrl']

ID
158023                   lionel-messi//
20801     c-ronaldo-dos-santos-aveiro//
200389                      jan-oblak//
192985                kevin-de-bruyne//
190871      neymar-da-silva-santos-jr//
                      ...              
247223                         ao-xia//
258760                      ben-hough//
252757                 ronan-mckinley//
243790                    zhenao-wang//
252520                      xiao-zhou//
Name: playerUrl, Length: 18979, dtype: object

In [11]:
# Step 4
fifa_df['playerUrl'] = fifa_df['playerUrl'].str.rstrip('/')
fifa_df['playerUrl']

ID
158023                   lionel-messi
20801     c-ronaldo-dos-santos-aveiro
200389                      jan-oblak
192985                kevin-de-bruyne
190871      neymar-da-silva-santos-jr
                     ...             
247223                         ao-xia
258760                      ben-hough
252757                 ronan-mckinley
243790                    zhenao-wang
252520                      xiao-zhou
Name: playerUrl, Length: 18979, dtype: object

In [12]:
# Step 5
fifa_df['playerUrl'] = fifa_df['playerUrl'].str.title()
fifa_df['playerUrl']

ID
158023                   Lionel-Messi
20801     C-Ronaldo-Dos-Santos-Aveiro
200389                      Jan-Oblak
192985                Kevin-De-Bruyne
190871      Neymar-Da-Silva-Santos-Jr
                     ...             
247223                         Ao-Xia
258760                      Ben-Hough
252757                 Ronan-Mckinley
243790                    Zhenao-Wang
252520                      Xiao-Zhou
Name: playerUrl, Length: 18979, dtype: object

In [13]:
# Step 6
fifa_df['playerUrl'].str.split('-', expand = True)

0         1       2       3       4     5
ID                                                    
158023  Lionel     Messi    None    None    None  None
20801        C   Ronaldo     Dos  Santos  Aveiro  None
200389     Jan     Oblak    None    None    None  None
192985   Kevin        De  Bruyne    None    None  None
190871  Neymar        Da   Silva  Santos      Jr  None
...        ...       ...     ...     ...     ...   ...
247223      Ao       Xia    None    None    None  None
258760     Ben     Hough    None    None    None  None
252757   Ronan  Mckinley    None    None    None  None
243790  Zhenao      Wang    None    None    None  None
252520    Xiao      Zhou    None    None    None  None

[18979 rows x 6 columns]

In [14]:
# Step 7
fifa_df[['name1','name2','name3','name4','name5','name6']] = fifa_df['playerUrl'].str.split('-', expand = True)
fifa_df.head()

Name                      LongName  \
ID                                                        
158023           L. Messi                  Lionel Messi   
20801   Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
200389           J. Oblak                     Jan Oblak   
192985       K. De Bruyne               Kevin De Bruyne   
190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                                photoUrl  \
ID                                                         
158023  https://cdn.sofifa.com/players/158/023/21_60.png   
20801   https://cdn.sofifa.com/players/020/801/21_60.png   
200389  https://cdn.sofifa.com/players/200/389/21_60.png   
192985  https://cdn.sofifa.com/players/192/985/21_60.png   
190871  https://cdn.sofifa.com/players/190/871/21_60.png   

                          playerUrl Nationality  Age  ↓OVA  POT  \
ID                                                                
158023                 Lionel-Messi   Argentina   33    93   93   
20801   C-Ronaldo-Dos-Santos-Aveiro    Portugal   35    92   92   
200389                    Jan-Oblak    Slovenia   27    91   93   
192985              Kevin-De-Bruyne     Belgium   29    91   91   
190871    Neymar-Da-Silva-Santos-Jr      Brazil   28    91   91   

                               Club     Contract  ... DRI DEF PHY Hits  \
ID                                                ...                    
158023         \n\n\n\nFC Barcelona  2004 ~ 2021  ...  95  38  65  771   
20801              \n\n\n\nJuventus  2018 ~ 2022  ...  89  35  77  562   
200389      \n\n\n\nAtlético Madrid  2014 ~ 2023  ...  90  52  90  150   
192985      \n\n\n\nManchester City  2015 ~ 2023  ...  88  64  78  207   
190871  \n\n\n\nParis Saint-Germain  2017 ~ 2022  ...  94  36  59  595   

         name1    name2   name3   name4   name5 name6  
ID                                                     
158023  Lionel    Messi    None    None    None  None  
20801        C  Ronaldo     Dos  Santos  Aveiro  None  
200389     Jan    Oblak    None    None    None  None  
192985   Kevin       De  Bruyne    None    None  None  
190871  Neymar       Da   Silva  Santos      Jr  None  

[5 rows x 82 columns]

In [15]:
# Step 8
fifa_df['Full Name'] = fifa_df[fifa_df.columns[76:]].apply(lambda x: ' '.join(x.dropna().astype(str)), axis = 1)
fifa_df.head()

Name                      LongName  \
ID                                                        
158023           L. Messi                  Lionel Messi   
20801   Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
200389           J. Oblak                     Jan Oblak   
192985       K. De Bruyne               Kevin De Bruyne   
190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                                photoUrl  \
ID                                                         
158023  https://cdn.sofifa.com/players/158/023/21_60.png   
20801   https://cdn.sofifa.com/players/020/801/21_60.png   
200389  https://cdn.sofifa.com/players/200/389/21_60.png   
192985  https://cdn.sofifa.com/players/192/985/21_60.png   
190871  https://cdn.sofifa.com/players/190/871/21_60.png   

                          playerUrl Nationality  Age  ↓OVA  POT  \
ID                                                                
158023                 Lionel-Messi   Argentina   33    93   93   
20801   C-Ronaldo-Dos-Santos-Aveiro    Portugal   35    92   92   
200389                    Jan-Oblak    Slovenia   27    91   93   
192985              Kevin-De-Bruyne     Belgium   29    91   91   
190871    Neymar-Da-Silva-Santos-Jr      Brazil   28    91   91   

                               Club     Contract  ... DEF PHY Hits   name1  \
ID                                                ...                        
158023         \n\n\n\nFC Barcelona  2004 ~ 2021  ...  38  65  771  Lionel   
20801              \n\n\n\nJuventus  2018 ~ 2022  ...  35  77  562       C   
200389      \n\n\n\nAtlético Madrid  2014 ~ 2023  ...  52  90  150     Jan   
192985      \n\n\n\nManchester City  2015 ~ 2023  ...  64  78  207   Kevin   
190871  \n\n\n\nParis Saint-Germain  2017 ~ 2022  ...  36  59  595  Neymar   

          name2   name3   name4   name5 name6                    Full Name  
ID                                                                          
158023    Messi    None    None    None  None                 Lionel Messi  
20801   Ronaldo     Dos  Santos  Aveiro  None  C Ronaldo Dos Santos Aveiro  
200389    Oblak    None    None    None  None                    Jan Oblak  
192985       De  Bruyne    None    None  None              Kevin De Bruyne  
190871       Da   Silva  Santos      Jr  None    Neymar Da Silva Santos Jr  

[5 rows x 83 columns]

In [16]:
# Step 9
drop_cols = ['Name',
             'LongName',
             'photoUrl',
             'playerUrl',
             'name1',
             'name2',
             'name3',
             'name4',
             'name5',
             'name6']

fifa_df = fifa_df.drop(drop_cols, axis = 1)
fifa_df.head()

Nationality  Age  ↓OVA  POT                         Club     Contract  \
ID                                                                             
158023   Argentina   33    93   93         \n\n\n\nFC Barcelona  2004 ~ 2021   
20801     Portugal   35    92   92             \n\n\n\nJuventus  2018 ~ 2022   
200389    Slovenia   27    91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023   
192985     Belgium   29    91   91      \n\n\n\nManchester City  2015 ~ 2023   
190871      Brazil   28    91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022   

         Positions Height Weight Preferred Foot  ...     D/W   IR PAC SHO PAS  \
ID                                               ...                            
158023  RW, ST, CF  170cm   72kg           Left  ...     Low  5 ★  85  92  91   
20801       ST, LW  187cm   83kg          Right  ...     Low  5 ★  89  93  81   
200389          GK  188cm   87kg          Right  ...  Medium  3 ★  87  92  78   
192985     CAM, CM  181cm   70kg          Right  ...    High  4 ★  76  86  93   
190871     LW, CAM  175cm   68kg          Right  ...  Medium  5 ★  91  85  86   

       DRI DEF  PHY  Hits                    Full Name  
ID                                                      
158023  95  38   65   771                 Lionel Messi  
20801   89  35   77   562  C Ronaldo Dos Santos Aveiro  
200389  90  52   90   150                    Jan Oblak  
192985  88  64   78   207              Kevin De Bruyne  
190871  94  36   59   595    Neymar Da Silva Santos Jr  

[5 rows x 73 columns]

In [17]:
# Look at the column data type
fifa_df['Full Name'].info()

<class 'pandas.core.series.Series'>
Int64Index: 18979 entries, 158023 to 252520
Series name: Full Name
Non-Null Count  Dtype 
--------------  ----- 
18979 non-null  object
dtypes: object(1)
memory usage: 296.5+ KB


### Process:
- The first unwanted 32 characters were removed from the 'playerUrl' column using `str.slice()` and `str.lstrip()`.
- The numbers behind the names in 'playerUrl' column was replaced with empty space and also removed the last two characters using `str.rstrip()`.
- The names left in the 'playerUrl' column were formatted to title case and then split with the `'-'` into six columns named; 'name1','name2','name3','name4','name5' and 'name6'.
- The six named columns 'name1','name2','name3','name4','name5' and 'name6' were joined into a single column and renamed to 'Full Name'.
- 'Name', 'LongName', 'photoUrl', 'playerUrl', 'name1','name2','name3','name4','name5' and 'name6' columns were dropped from the fifa_df.
- The 'Full Name' column data type was object which is correct.

## Nationality, Age, Club, ↓OVA, POT and BOV
Now, lets move on to the next columns listed above.

In [18]:
# Inspect Nationality, Age, Club, ↓OVA, POT and 'BOV' columns
fifa_df[['Nationality', 'Age', 'Club', '↓OVA', 'POT', 'BOV']].head()

Nationality  Age                         Club  ↓OVA  POT  BOV
ID                                                                  
158023   Argentina   33         \n\n\n\nFC Barcelona    93   93   93
20801     Portugal   35             \n\n\n\nJuventus    92   92   92
200389    Slovenia   27      \n\n\n\nAtlético Madrid    91   93   91
192985     Belgium   29      \n\n\n\nManchester City    91   91   91
190871      Brazil   28  \n\n\n\nParis Saint-Germain    91   91   91

- 'Nationality' and 'Age' column had no missing entries.
- 'Club' column consist of several unwanted character issues wrapped around the club names.
- Since '↓OVA', 'POT' and 'BOV' had values ranges from 1-99 which was all about players rating on a scale of 1% to 100%.
- It would be best to convert the values in '↓OVA', 'POT' and 'BOV' columns to percentage (%).

In [19]:
# Remove unwanted characters using lstrip
fifa_df['Club'] = fifa_df['Club'].str.lstrip('\n')
fifa_df['Club'].head()

ID
158023           FC Barcelona
20801                Juventus
200389        Atlético Madrid
192985        Manchester City
190871    Paris Saint-Germain
Name: Club, dtype: object

In [20]:
# Rename ↓OVA column to OVA
fifa_df = fifa_df.rename(columns = {'↓OVA' : 'OVA'})

# Convert 'OVA', 'POT' and 'BOV' column values to percent
fifa_df[['OVA', 'POT', 'BOV']] = fifa_df[['OVA', 'POT', 'BOV']].astype(str) + '%'
fifa_df[['OVA', 'POT', 'BOV']].head()

OVA  POT  BOV
ID                   
158023  93%  93%  93%
20801   92%  92%  92%
200389  91%  93%  91%
192985  91%  91%  91%
190871  91%  91%  91%

In [21]:
# Look at each column data type
fifa_df[['Nationality', 'Age', 'Club', 'OVA', 'POT', 'BOV']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Nationality  18979 non-null  object
 1   Age          18979 non-null  int64 
 2   Club         18979 non-null  object
 3   OVA          18979 non-null  object
 4   POT          18979 non-null  object
 5   BOV          18979 non-null  object
dtypes: int64(1), object(5)
memory usage: 1.0+ MB


- 'Nationality' and 'Club' column data type was object, while 'Age' column data type was integer which are correct.
- Unwanted characters in 'Club' column was removed using `str.lstrip()`.
-  ↓OVA column was renamed to 'OVA'.
- 'OVA', 'POT' and 'BOV' column values was converted percentage by adding `%` to each column values.
- '↓OVA', 'POT', 'BOV' column data type was object due to `%` added to the data entries.

## Contract

In [22]:
# Inspect Contract column
fifa_df['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

- 'Contract' column was inspected and observed that it entails three different data entry values such as '2004 ~ 2021', 'June 30, 2021 On Loan' and 'Free' using `fifa_df.unique()`.
- Since we have identified the issues associated with this column. Then, lets get our hands on the neccessary process.

In [23]:
# Define a function to change column values
def to_status(x):
    if 'On Loan' in x:
        x = 'On Loan'
        return x
    elif '~' in x:
        x = 'Active'
        return x
    elif 'Free' in x:
        x = 'Free'
        return x

fifa_df['Contract'] = fifa_df['Contract'].apply(to_status).astype('category')
fifa_df['Contract'].unique()

['Active', 'On Loan', 'Free']
Categories (3, object): ['Active', 'Free', 'On Loan']

In [24]:
# Rename the column to Contract Status
fifa_df = fifa_df.rename(columns = {'Contract' : 'Contract Status'})
fifa_df['Contract Status'].head()

ID
158023    Active
20801     Active
200389    Active
192985    Active
190871    Active
Name: Contract Status, dtype: category
Categories (3, object): ['Active', 'Free', 'On Loan']

### Process:
- A function was defined to change the row values from '2004 ~ 2021', 'On Loan' and 'Free' to 'Active', 'On Loan' and 'Free'
- 'Contract' column was renamed to 'Contract Status'.
- Also, the data type was changed to `category`.

## Positions and Best Position

In [25]:
# Inspect Positions and Best Position column
fifa_df[['Positions', 'Best Position']].head()

Positions Best Position
ID                              
158023  RW, ST, CF            RW
20801       ST, LW            ST
200389          GK            GK
192985     CAM, CM           CAM
190871     LW, CAM            LW

In [26]:
# Drop Positions column
fifa_df = fifa_df.drop(['Positions'], axis = 1)
fifa_df.head()

Nationality  Age  OVA  POT                 Club Contract Status Height  \
ID                                                                              
158023   Argentina   33  93%  93%         FC Barcelona          Active  170cm   
20801     Portugal   35  92%  92%             Juventus          Active  187cm   
200389    Slovenia   27  91%  93%      Atlético Madrid          Active  188cm   
192985     Belgium   29  91%  91%      Manchester City          Active  181cm   
190871      Brazil   28  91%  91%  Paris Saint-Germain          Active  175cm   

       Weight Preferred Foot  BOV  ...     D/W   IR PAC SHO PAS DRI  DEF  PHY  \
ID                                 ...                                          
158023   72kg           Left  93%  ...     Low  5 ★  85  92  91  95   38   65   
20801    83kg          Right  92%  ...     Low  5 ★  89  93  81  89   35   77   
200389   87kg          Right  91%  ...  Medium  3 ★  87  92  78  90   52   90   
192985   70kg          Right  91%  ...    High  4 ★  76  86  93  88   64   78   
190871   68kg          Right  91%  ...  Medium  5 ★  91  85  86  94   36   59   

        Hits                    Full Name  
ID                                         
158023   771                 Lionel Messi  
20801    562  C Ronaldo Dos Santos Aveiro  
200389   150                    Jan Oblak  
192985   207              Kevin De Bruyne  
190871   595    Neymar Da Silva Santos Jr  

[5 rows x 72 columns]

In [27]:
# Look at the column data type
fifa_df['Best Position'].info()

<class 'pandas.core.series.Series'>
Int64Index: 18979 entries, 158023 to 252520
Series name: Best Position
Non-Null Count  Dtype 
--------------  ----- 
18979 non-null  object
dtypes: object(1)
memory usage: 296.5+ KB


### Process:
- 'Positions' and 'Best Position' column was inspected and it was observed that the entry values in both looks similar.
- Then, it was realize that 'Positions' column had one and also extra position for each player that had more than one position.
- 'Positions' column was dropped from the fifa_df. Because, it was already in 'Best Position' column but only one specific position to a player.
- 'Best Position' column data type was object which is correct.

## Height, Weight and Preferred Foot

In [28]:
# Inspect Height, Weight, Preferred Foot and BOV columns
for column in fifa_df[['Height', 'Weight', 'Preferred Foot']]:
    values = fifa_df[column].unique()
    print(f'{column}-{values}.\n')

Height-['170cm' '187cm' '188cm' '181cm' '175cm' '184cm' '191cm' '178cm' '193cm'
 '185cm' '199cm' '173cm' '168cm' '176cm' '177cm' '183cm' '180cm' '189cm'
 '179cm' '195cm' '172cm' '182cm' '186cm' '192cm' '165cm' '194cm' '167cm'
 '196cm' '163cm' '190cm' '174cm' '169cm' '171cm' '197cm' '200cm' '166cm'
 '6\'2"' '164cm' '198cm' '6\'3"' '6\'5"' '5\'11"' '6\'4"' '6\'1"' '6\'0"'
 '5\'10"' '5\'9"' '5\'6"' '5\'7"' '5\'4"' '201cm' '158cm' '162cm' '161cm'
 '160cm' '203cm' '157cm' '156cm' '202cm' '159cm' '206cm' '155cm'].

Weight-['72kg' '83kg' '87kg' '70kg' '68kg' '80kg' '71kg' '91kg' '73kg' '85kg'
 '92kg' '69kg' '84kg' '96kg' '81kg' '82kg' '75kg' '86kg' '89kg' '74kg'
 '76kg' '64kg' '78kg' '90kg' '66kg' '60kg' '94kg' '79kg' '67kg' '65kg'
 '59kg' '61kg' '93kg' '88kg' '97kg' '77kg' '62kg' '63kg' '95kg' '100kg'
 '58kg' '183lbs' '179lbs' '172lbs' '196lbs' '176lbs' '185lbs' '170lbs'
 '203lbs' '168lbs' '161lbs' '146lbs' '130lbs' '190lbs' '174lbs' '148lbs'
 '165lbs' '159lbs' '192lbs' '181lbs' '139lbs' '15

- Loop through using `fifa_df.unique()` to inspect 'Height', 'Weight' and 'Preferred Foot' column.
- 'Height' column had two different data entry values in '178cm' and '5'10"(feet & inches).
- 'Weight' column had two different data entry values in '75kg' and '110lbs'.
- 'Preferred Foot' column also had two different data entry values in 'Left' and 'Right'.
- Having identified these, both 'Height' and 'Weight' column values needs to be in a specific measure to be able to perform further analysis, while 'Preferred Foot' column values is okay as it is.

In [29]:
# Define a function to convert values to kg
def all_to_kg(x):
    if 'kg' in x:
        x = x.strip('kg')
        return x
    elif 'lbs' in x:
        x = x.strip('lbs')
        Weight = round((float(x) * 0.45359237), 2)
        return Weight

fifa_df['Weight'] = fifa_df['Weight'].apply(all_to_kg).astype('int64')
fifa_df['Weight'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52], dtype=int64)

In [30]:
# Define a function to convert values to cm
def all_to_cm(x):
    if "'" and '"' in x:
        feet, inches = x.split("'")
        total_inches = int(feet) * 12 + int(inches[:-1])
        height_cm = total_inches * 2.54
    elif 'cm' in x:
        height_cm = int(x[:-2])
        total_inches = round(height_cm / 2.54)
    else:
        feet, inches = x.split("'")
        total_inches = int(feet) * 12 + int(inches[:-1])
        height_cm = total_inches * 2.54
    return round(height_cm, 2)

fifa_df['Height'] = fifa_df['Height'].apply(all_to_cm).astype('int64')
fifa_df['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 162,
       201, 158, 161, 160, 203, 157, 156, 202, 159, 206, 155], dtype=int64)

In [31]:
# Rename 'Height' and 'Weight' column by adding suffix of 'cm & kg'
fifa_df = fifa_df.rename(columns = {'Height' : 'Height (cm)', 'Weight' : 'Weight (kg)'})
fifa_df[['Height (cm)', 'Weight (kg)']].head(2)

Height (cm)  Weight (kg)
ID                              
158023          170           72
20801           187           83

In [32]:
# Look at the column data type
fifa_df[['Height (cm)', 'Weight (kg)', 'Preferred Foot']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Height (cm)     18979 non-null  int64 
 1   Weight (kg)     18979 non-null  int64 
 2   Preferred Foot  18979 non-null  object
dtypes: int64(2), object(1)
memory usage: 593.1+ KB


### Process:
- A function was defined to remove 'kg and lbs' in 'Weight' column and convert all 'lbs' to 'kg' by multiplying it by (0.45359237) per kg and also change the data type to `int64` in the process.
- Another function was also defined to split, remove and convert 'cm and feet & inches' to 'cm' and change the data type to `int64` along the process.
- 'Height' and 'Weight' column was renamed to 'Height (cm)' and 'Weight (kg)'
- 'Preferred Foot' column was good as it is and left untouched.

## Joined and Loan Date End

In [33]:
# Inspect Joined and Loan Date End column
fifa_df[['Joined', 'Loan Date End']].sample(5)

Joined Loan Date End
ID                             
257967  17-Jul-20           NaN
200841   7-Aug-19           NaN
235710   1-Jul-13           NaN
237965   5-Jan-17           NaN
232576  13-Aug-19           NaN

In [34]:
# Step 1
fifa_df[['Joined', 'Loan Date End']].isnull().sum()

Joined               0
Loan Date End    17966
dtype: int64

In [35]:
# Step 2
fifa_df = fifa_df.fillna(value = {'Loan Date End' : 'Null'})
fifa_df['Loan Date End'].sample(5)

ID
257398         Null
237540         Null
225696    30-Jun-21
257175         Null
240257         Null
Name: Loan Date End, dtype: object

In [36]:
# Step 3
fifa_df['Joined'] = pd.to_datetime(fifa_df['Joined'])
fifa_df['Joined'].head()

ID
158023   2004-07-01
20801    2018-07-10
200389   2014-07-16
192985   2015-08-30
190871   2017-08-03
Name: Joined, dtype: datetime64[ns]

In [37]:
# Look at the column type
fifa_df[['Joined', 'Loan Date End']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Joined         18979 non-null  datetime64[ns]
 1   Loan Date End  18979 non-null  object        
dtypes: datetime64[ns](1), object(1)
memory usage: 444.8+ KB


### Process:
- 'Joined' and 'Loan Date End' column was inspected to see if they had data entry issues.
- 'Joined' column had no empty or data entry issues.
- 'Loan Date End' column had 17966 empty rows out of 18979 rows. This was understood as the total number of players 'On Loan' account for the non-null entry rows.
- 'Loan Date End' empty rows was filled with 'Null'. 
- The 'Joined' column data type was object and was changed to datetime data type. While the 'Loan Date End' column was object due to having 'Null' in some of the row entries.

## Value, Wage, Release Clause and Hits

In [38]:
# Inspect Value, Wage, Release Clause and Hits columns
for column in fifa_df[['Value', 'Wage', 'Release Clause', 'Hits']]:
    values = fifa_df[column].unique()
    print(f'{column}-{values}.\n')

Value-['€103.5M' '€63M' '€120M' '€129M' '€132M' '€111M' '€120.5M' '€102M'
 '€185.5M' '€110M' '€113M' '€90.5M' '€82M' '€17.5M' '€83.5M' '€33.5M'
 '€114.5M' '€78M' '€103M' '€109M' '€92M' '€10M' '€76.5M' '€89.5M' '€87.5M'
 '€79.5M' '€124M' '€114M' '€95M' '€92.5M' '€105.5M' '€88.5M' '€85M'
 '€81.5M' '€26M' '€21M' '€56M' '€67.5M' '€53M' '€36.5M' '€51M' '€65.5M'
 '€46.5M' '€61.5M' '€72.5M' '€77.5M' '€43.5M' '€32.5M' '€36M' '€32M'
 '€54M' '€49.5M' '€57M' '€66.5M' '€74.5M' '€71.5M' '€121M' '€99M' '€67M'
 '€86.5M' '€93.5M' '€70M' '€62M' '€66M' '€58M' '€44M' '€81M' '€37M'
 '€14.5M' '€46M' '€47.5M' '€52.5M' '€54.5M' '€34.5M' '€57.5M' '€51.5M'
 '€44.5M' '€55M' '€48M' '€60.5M' '€63.5M' '€61M' '€29M' '€58.5M' '€55.5M'
 '€42M' '€40.5M' '€43M' '€45.5M' '€34M' '€26.5M' '€42.5M' '€35.5M' '€45M'
 '€41.5M' '€40M' '€11M' '€13.5M' '€29.5M' '€27M' '€15.5M' '€38.5M' '€52M'
 '€33M' '€19M' '€73.5M' '€38M' '€35M' '€47M' '€24M' '€30.5M' '€18M' '€28M'
 '€25.5M' '€25M' '€31M' '€23.5M' '€30M' '€31.5M' '€22.5M' '€28.

- Loop through using `fifa_df.unique()` to inspect 'Value', 'Wage' and 'Release Clause' column.
- The aforementioned columns had different data entries with prefixes and suffixes such as '€' for euro currency, 'K' for thousands, 'M' for millions.
- 'Hits' column also had data entry issue such as suffixes of 'K' for thousands as well as blank/empty 'nan' rows.

In [39]:
# Step 1
fifa_df = fifa_df.fillna(value = {'Hits' : '0'})
fifa_df['Hits'].sample(5)

ID
251809    91
238593    14
224670     0
252087     4
222277     1
Name: Hits, dtype: object

In [40]:
# Step 2
for column in fifa_df[['Value', 'Wage', 'Release Clause']]:
    fifa_df[column] = fifa_df[column].str.strip('€')

fifa_df[['Value', 'Wage', 'Release Clause', 'Hits']].sample(5)

Value Wage Release Clause Hits
ID                                   
233017  1.3M   2K           979K    3
203886    3M  19K           5.7M    8
209660  2.3M   9K           3.9M   15
190666  3.5M  21K           8.5M    3
201122  2.1M  24K           5.1M   11

In [41]:
# Step 3
for column in fifa_df[['Value', 'Wage', 'Release Clause', 'Hits']]:
    fifa_df[column] = fifa_df[column].replace({'K':'*1e3', 'M':'*1e6'}, regex = True).map(pd.eval).astype('int64')
    
fifa_df[['Value', 'Wage', 'Release Clause', 'Hits']].sample(5)

Value   Wage  Release Clause  Hits
ID                                          
207734  1800000   7000         2100000    14
142966  1000000  10000         2000000     8
223264   525000   2000          639000     3
198534   725000   3000          656000     1
243445  1100000   2000         1100000     7

In [42]:
# Step 4
fifa_df = fifa_df.rename(columns = {'Value' : 'Value (€)', 'Wage' : 'Wage (€)', 'Release Clause' : 'Release Clause (€)'})
fifa_df[['Value (€)', 'Wage (€)', 'Release Clause (€)']].head(2)

Value (€)  Wage (€)  Release Clause (€)
ID                                             
158023  103500000    560000           138400000
20801    63000000    220000            75900000

In [43]:
# Look at the column type
fifa_df[['Value (€)', 'Wage (€)', 'Release Clause (€)', 'Hits']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   Value (€)           18979 non-null  int64
 1   Wage (€)            18979 non-null  int64
 2   Release Clause (€)  18979 non-null  int64
 3   Hits                18979 non-null  int64
dtypes: int64(4)
memory usage: 741.4 KB


### Process:
- 'Hits' column empty rows 'nan' was filled with zero (0) for players that had no hits.
- Loop through 'Value', 'Wage', 'Release Clause' and 'Hits' columns to remove '€' using `str.strip()`.
- Loop through 'Value', 'Wage', 'Release Clause' and 'Hits' columns again to replace 'K' & 'M' with '*1e3' &'*1e6' for pandas to evaluate 'K' & 'M' found in any of the column values to thousands & millions and then changed the data type to `int64` along the process.
- 'Value', 'Wage' and 'Release Clause' column was renamed to 'Value (€)', 'Wage (€)' and 'Release Clause (€)'

### Attacking, Crossing, Finishing, Heading Accuracy, Short Passing, Volleys, Skill, Dribbling, Curve, FK Accuracy, Long Passing, Ball Control, Movement, Acceleration, Sprint Speed, Agility, Reactions, Balance, Power,  Shot Power, Jumping, Stamina, Strength, Long Shots, Mentality, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Defending, Marking, Standing Tackle, Sliding Tackle, Goalkeeping, GK Diving, GK Handling, GK Kicking, GK Positioning, GK Reflexes, Total Stats, Base Stats, A/W, D/W, PAC, SHO, PAS, DRI, DEF and lastly PHY.

In [44]:
# Inspect the listed columns above
fifa_df[['Attacking', 'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve',
         'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility', 'Reactions',
         'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots', 'Mentality', 'Aggression',
         'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
         'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning', 'GK Reflexes',
         'Total Stats', 'Base Stats', 'A/W', 'D/W', 'PAC', 'SHO', 'PAS', 'DRI', 'DEF', 'PHY']].sample(10)

Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  \
ID                                                                        
190438        219        29         26                66             61   
228790        252        68         36                53             64   
254847        181        30         18                61             45   
186882        102        16         11                18             43   
230860        297        64         57                55             72   
190772        270        49         35                68             74   
243320        237        48         49                43             58   
235534        238        57         32                53             61   
254298         62        15         10                12             19   
187077        312        64         65                61             62   

        Volleys  Skill  Dribbling  Curve  FK Accuracy  ...  Total Stats  \
ID                                                     ...                
190438       37    236         24     34           62  ...         1519   
228790       31    268         65     44           36  ...         1696   
254847       27    161         25     26           23  ...         1258   
186882       14    110         16     17           21  ...         1192   
230860       49    336         67     64           66  ...         1912   
190772       44    262         58     34           35  ...         1719   
243320       39    254         55     45           42  ...         1574   
235534       35    271         61     53           44  ...         1580   
254298        6     73         15      9           10  ...          820   
187077       60    317         65     62           66  ...         1771   

        Base Stats     A/W     D/W  PAC  SHO  PAS  DRI  DEF  PHY  
ID                                                                
190438         311  Medium  Medium   42   43   51   37   66   72  
228790         374    High  Medium   79   37   59   67   62   70  
254847         279  Medium  Medium   59   22   37   36   54   71  
186882         386  Medium  Medium   72   66   65   74   37   72  
230860         409  Medium  Medium   74   61   68   69   66   71  
190772         367     Low    High   60   42   62   60   70   73  
243320         337  Medium  Medium   69   50   54   58   50   56  
235534         329  Medium  Medium   66   41   54   62   57   49  
254298         282  Medium  Medium   52   50   55   53   30   42  
187077         366    High    High   75   64   62   69   37   59  

[10 rows x 51 columns]

In [45]:
# Check for missing values
fifa_df[['Attacking', 'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve',
         'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility', 'Reactions',
         'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots', 'Mentality', 'Aggression',
         'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
         'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning', 'GK Reflexes',
         'Total Stats', 'Base Stats', 'A/W', 'D/W', 'PAC', 'SHO', 'PAS', 'DRI', 'DEF', 'PHY']].isnull().sum()

Attacking           0
Crossing            0
Finishing           0
Heading Accuracy    0
Short Passing       0
Volleys             0
Skill               0
Dribbling           0
Curve               0
FK Accuracy         0
Long Passing        0
Ball Control        0
Movement            0
Acceleration        0
Sprint Speed        0
Agility             0
Reactions           0
Balance             0
Power               0
Shot Power          0
Jumping             0
Stamina             0
Strength            0
Long Shots          0
Mentality           0
Aggression          0
Interceptions       0
Positioning         0
Vision              0
Penalties           0
Composure           0
Defending           0
Marking             0
Standing Tackle     0
Sliding Tackle      0
Goalkeeping         0
GK Diving           0
GK Handling         0
GK Kicking          0
GK Positioning      0
GK Reflexes         0
Total Stats         0
Base Stats          0
A/W                 0
D/W                 0
PAC       

In [46]:
# look at the column type
fifa_df[['Attacking', 'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve',
         'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility', 'Reactions',
         'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots', 'Mentality', 'Aggression',
         'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
         'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning', 'GK Reflexes',
         'Total Stats', 'Base Stats', 'A/W', 'D/W', 'PAC', 'SHO', 'PAS', 'DRI', 'DEF', 'PHY']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 51 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Attacking         18979 non-null  int64 
 1   Crossing          18979 non-null  int64 
 2   Finishing         18979 non-null  int64 
 3   Heading Accuracy  18979 non-null  int64 
 4   Short Passing     18979 non-null  int64 
 5   Volleys           18979 non-null  int64 
 6   Skill             18979 non-null  int64 
 7   Dribbling         18979 non-null  int64 
 8   Curve             18979 non-null  int64 
 9   FK Accuracy       18979 non-null  int64 
 10  Long Passing      18979 non-null  int64 
 11  Ball Control      18979 non-null  int64 
 12  Movement          18979 non-null  int64 
 13  Acceleration      18979 non-null  int64 
 14  Sprint Speed      18979 non-null  int64 
 15  Agility           18979 non-null  int64 
 16  Reactions         18979 non-null  int64 
 17  Balanc

### Process:
- 'Attacking', 'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots', 'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle', 'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats', 'Base Stats', 'A/W', 'D/W', 'PAC', 'SHO', 'PAS', 'DRI', 'DEF' and 'PHY' columns were inspected and all had no empty rows or inconsistencies issues among the data entries as well as the data type.

### W/F, SM and IR

In [47]:
# Inspect W/F, SM and IR columns
fifa_df[['W/F', 'SM', 'IR']].head()

W/F  SM   IR
ID                  
158023  4 ★  4★  5 ★
20801   4 ★  5★  5 ★
200389  3 ★  1★  3 ★
192985  5 ★  4★  4 ★
190871  5 ★  5★  5 ★

In [48]:
# Loop through to remove ★ from each of the column row values
for column in fifa_df[['W/F', 'SM', 'IR']]:
    fifa_df[column] = fifa_df[column].str.rstrip('★').astype('int64')
    
fifa_df[['W/F', 'SM', 'IR']].head()

W/F  SM  IR
ID                 
158023    4   4   5
20801     4   5   5
200389    3   1   3
192985    5   4   4
190871    5   5   5

In [49]:
# Look at the column type
fifa_df[['W/F', 'SM', 'IR']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   W/F     18979 non-null  int64
 1   SM      18979 non-null  int64
 2   IR      18979 non-null  int64
dtypes: int64(3)
memory usage: 593.1 KB


### Process:
- 'W/F', 'SM' and 'IR' columns had value ratings on a scale of 1-5 which also had special character '★' as suffixes to each of the scale in each of the column scale entries.
- Loop through the columns, the special character '★' was removed using `str.rstrip()`.
- The columns data type was object and changed to `int64` in the process of removing '★' .

In [50]:
# Obtain information about the column data types
fifa_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 158023 to 252520
Data columns (total 72 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Nationality         18979 non-null  object        
 1   Age                 18979 non-null  int64         
 2   OVA                 18979 non-null  object        
 3   POT                 18979 non-null  object        
 4   Club                18979 non-null  object        
 5   Contract Status     18979 non-null  category      
 6   Height (cm)         18979 non-null  int64         
 7   Weight (kg)         18979 non-null  int64         
 8   Preferred Foot      18979 non-null  object        
 9   BOV                 18979 non-null  object        
 10  Best Position       18979 non-null  object        
 11  Joined              18979 non-null  datetime64[ns]
 12  Loan Date End       18979 non-null  object        
 13  Value (€)           18979 non-null  int6

In [51]:
fifa_df.to_csv('../DataChallenge/fifa21_cleaned_data.csv')

###### Cleaned By: `Toheeb Olamilekan BABATUNDE`